### ISSUE TITLE
Wildcard pattern is matching files incorrectly.

### ISSUE THREAD

**Issue by jeffkl** [dotnet/corefx#27779]
Some MSBuild tests started failing against the latest .NET CLI preview because file pattern matching is returning unexpected results on Linux and Mac.  

# Repro:

1. Create two files in the program output directory, `Foo.cs` and `Foo2.cs`
2. `dotnet run` the following:
```C#
using System;
using System.IO;

namespace wildcards
{
    class Program
    {
        static void Main(string[] args)
        {
            foreach (string path in Directory.EnumerateFiles(".", "????.cs"))
            {
                Console.WriteLine(path);
            }
        }
    }
}
```

# Expected Result
Program should print out the path to `Foo2.cs` since its file name has four characters that match the four questions marks.
```
/Users/UserA/Foo/bin/Debug/netcoreapp2.0/Foo2.cs
```

# Actual Result
Program prints out the path to `Foo.cs` and `Foo2.cs` since `Foo2.cs` is the only file that matches `????.cs`

```
/Users/UserA/Foo/bin/Debug/netcoreapp2.0/Foo.cs
/Users/UserA/Foo/bin/Debug/netcoreapp2.0/Foo2.cs
```

# dotnet --info
```
.NET Command Line Tools (2.1.300-preview2-008293)

Product Information:
 Version:            2.1.300-preview2-008293
 Commit SHA-1 hash:  68922e2a51

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  16.04
 OS Platform: Linux
 RID:         ubuntu.16.04-x64
 Base Path:   /home/jeff/msbuild/artifacts/.dotnet/2.1.300-preview2-008293/sdk/2.1.300-preview2-008293/

Host (useful for support):
  Version: 2.1.0-preview2-26227-01
  Commit:  86d1f92013

.NET Core SDKs installed:
  2.1.300-preview2-008293 [/home/jeff/msbuild/artifacts/.dotnet/2.1.300-preview2-008293/sdk]

The.NET Core runtimes installed:
  Microsoft.NETCore.App 2.1.0-preview2-26227-01 [/home/jeff/msbuild/artifacts/.dotnet/2.1.300-preview2-008293/shared]
  Microsoft.AspNetCore.App 2.1.0-preview2-30171 [/home/jeff/msbuild/artifacts/.dotnet/2.1.300-preview2-008293/shared]
  Microsoft.AspNetCore.All 2.1.0-preview2-30171 [/home/jeff/msbuild/artifacts/.dotnet/2.1.300-preview2-008293/shared]
```

**Comment by karelz** [comment]
@jeffkl do you know if it is a regression against 2.0?

**Comment by jeffkl** [comment]
We only saw it after we updated to `2.1.300-preview2-008293` so I don't think so.

**Comment by karelz** [comment]
@jeffkl which version did you update from? Earlier 2.1? 2.0? 1.x?

**Comment by jeffkl** [comment]
We noticed it in a PR from `2.1.300-preview2-00804` to `2.1.300-preview2-008293`.

https://github.com/Microsoft/msbuild/pull/2961/files#diff-9260067da6a2db86b9a5db30fa8ee2ae

**Comment by JeremyKuhne** [comment]
This was a bug fix. The wildcard results now match on Windows and Unix. `?` matches can collapse to a period as they do on Windows. (You can try this at a CMD prompt to see the behavior.) We have provided new options that allow you to use "simple" wildcard matching on all OSes as well.

https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System/IO/EnumerationOptions.cs

There are overloads in `Directory` and `DirectoryInfo` that allow you to use this new options class. "Simple" matching is the default when using this class. Note that skipping hidden and system files is also a default when using this class.

**Comment by jeffkl** [comment]
Okay we'll see if we can workaround this.  As of now its only affecting our ability to move to a newer CLI so the impact is minimal.

/cc @cdmihai

**Comment by cdmihai** [comment]
> skipping hidden and system files is also a default when using this class.

Is this requirement part of the general goal of making .net core behave the same as full framework? As in, are hidden and system files ignored on full framework (I guess I can also test this ...). I am asking to understand the extent to which full framework and .net core msbuild are affected. I expect build users would want everything, and don't care whether something is hidden or system.

**Comment by cdmihai** [comment]
Also, is there a blog post or something enumerating all the breaking changes? We could use that to understand the extent to which we need to react in msbuild.

**Comment by JeremyKuhne** [comment]
> are hidden and system files ignored on full framework

No, and they aren't ignored in the existing APIs. Existing APIs will return results as they always have.

The idea was that the new `EnumerationOptions` would have more user and cross-plat focused defaults. It matches simply, for example. Most people don't understand the weirdness around how `?` and `*` matching works on Windows. There is no "magic" in the simple matching- `?` is one char, and `*` is 0+.  Additionally, most don't want the `.vs`, `.git`, `$RECYCLE.BIN`, `Thumbs.db` etc. folders in results.

> is there a blog post or something enumerating all the breaking changes?

I'm working on a blog post. The key behavioral changes:

- We've made matching consistent with legacy behavior on non-Windows platforms
- We've fixed enumeration of unusual Windows files - you can now get `*Info` classes out successfully and use the methods on files that end in spaces, periods
- We do very little validation of paths up front as there is very little we can accurately predict (unblocking numerous xplat and some existing Windows scenarios)
    - We only check for embedded nulls, **no** other chars are rejected, including wildcards (as nulls are never supported and APIs almost universally take null terminated strings)
    - We don't check for "proper" colon placement
    - We don't check for segment length or total path length
    - We don't check for "proper" UNCs
    - We still throw for null or empty paths on all platforms or paths of all spaces on Windows (as they throw from GetFullPathName)
- We don't trim leading spaces on any paths on Windows anymore (we did for some, not others)
- We don't trim whitespace characters from the end of paths on Windows (such as nbsp)
- GetPathRoot, GetDirectoryName, etc. don't throw for empty anymore, they return null (like they do for null).
- GetPathRoot now works consistently with various Windows prefixes (e.g. `\\?\`, `\\.\`)

Key impacts:

- Exceptions for "bad" paths are thrown when used, not when normalizing (as we can't know what is "valid" without using them
- Cross plat scenarios are no longer blocked (mounting Unix volumes/shares on Windows for one example)
- Performance is significantly increased (we spend a huge amount of time checking invalid characters over and over)
- Behavior is more consistent when running code across platforms
